mitodl/open-discussions
# Show related posts on the post detail page

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #1748

#### What's this PR do?
Shows related posts on the post detail page

#### How should this be manually tested?
- Set `FEATURE_RELATED_POSTS_UI=True`
- Visit a post detail page
- Confirm that 4 related posts are shown in a sidebar with correct links, and that there seems to be a relationship between the similarity of those posts & how high they appear on the list

Additional testing:
- Make sure everything works as expected when that feature flag is set to false
- Make sure you can't see private channel posts in the related sidebar if your logged-in user is not a member, and make sure that you can see them in the opposite case

#### Any background context you want to provide?
[More Like This query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-mlt-query.html)

For this feature we're querying ES for related content based on a specific document (i.e.: the one you're looking at on a post detail page) rather than by a string or set of strings. The fields that we're matching on: `"plain_text", "post_title", "author_id", "channel_name"`. If a post has similar content in one or more of those fields, its score will be boosted in the "more like this" query results. It's obvious that posts with similar titles and similar text content are boosted, but we're also choosing to boost content with the same author and posted in the same channel in order to "fill out" the results in case there aren't any matches based on title/content.

A note about link posts: URL values cannot be partially matched in ES by default – they are stored as a single "word". This means that a link post with the URL "mit.edu/academic_calendar" and another with the URL "mit.edu/academics" will not match (in other words, they will not be considered related). The `post_link_url` field was left off of the "more like this" fields list for that reason. We are considering the possibility of indexing summary content that we get from embedly, which would give us something more meaningful to compare between link posts. Until then, they will only match based on their titles, authors, and channels.

#### Screenshots (if appropriate)
![ss 2019-02-01 at 13 50 50](https://user-images.githubusercontent.com/14932219/52143369-44b8ff00-2629-11e9-8473-289160c8f50e.png)
![ss 2019-02-01 at 13 51 34](https://user-images.githubusercontent.com/14932219/52143371-4682c280-2629-11e9-994e-2d69b2a1acfd.png)
